yarn-project/noir-contracts/src/contracts/uniswap_contract/src/main.nr

```diff
@@ -49,6 +49,96 @@ contract Uniswap {
     #[aztec(private)]
     fn constructor() {}
 
+    #[aztec(public)]
+    fn swap_public(
+        sender: AztecAddress,
+        input_asset_bridge: AztecAddress, 
+        input_amount: Field,
+        output_asset_bridge: AztecAddress,
+        // params for using the transfer approval
+        nonce_for_transfer_approval: Field,
+        // params for the swap
+        uniswap_fee_tier: Field,
+        minimum_output_amount: Field,
+        // params for the depositing output_asset back to Aztec
+        recipient: AztecAddress,
+        secret_hash_for_L1_to_l2_message: Field,
+        deadline_for_L1_to_l2_message: Field,
+        canceller_for_L1_to_L2_message: EthereumAddress,
+        caller_on_L1: EthereumAddress,
+        // nonce for someone to call swap on sender's behalf
+        nonce_for_swap_approval: Field,
+    ) -> Field {
+
+        if (sender.address != context.msg_sender()) {
+            // if someone else is calling on swap on sender's behalf, they need to have authorisation to do so:
+            let selector = compute_selector(
+                "swap_public((Field),(Field),Field,(Field),Field,Field,Field,(Field),Field,Field,(Field),(Field),Field)"
+            );
+            let message_field = compute_message_hash([
+                context.msg_sender(),
+                context.this_address(),
+                selector,
+                sender.address,
+                input_asset_bridge.address,
+                input_amount,
+                output_asset_bridge.address,
+                nonce_for_transfer_approval,
+                uniswap_fee_tier,
+                minimum_output_amount,
+                recipient.address,
+                secret_hash_for_L1_to_l2_message,
+                deadline_for_L1_to_l2_message,
+                canceller_for_L1_to_L2_message.address,
+                caller_on_L1.address,
+                nonce_for_swap_approval,
+            ]);
+            // this also emits a nullifier for the message
+            assert_valid_public_message_for(&mut context,sender.address,message_field);
+        }
+
+        let input_asset = AztecAddress::new(TokenBridge::at(input_asset_bridge.address).token(context));
+
+        // Transfer funds to this contract
+        Token::at(input_asset.address).transfer_public(
+            context,
+            sender.address,
+            context.this_address(),
+            input_amount,
+            nonce_for_transfer_approval,
```

Comment:
just thinking, for public calls does nonce still need to be send around? Could the account contract itself maintain it in state? then it an be infered?

Bare with me here as i might be horribly wrong

Reply:
Say you storage the nonce in the contract as a counter or whatever. What nonce would the uniswap contract read if you have done multiple approvals? If we expect only one approval to be active at a time you can do it quite easily, but in some cases, we might want multiple approvals within the same tx as well, then the ordering we approve them in suddenly becomes important as well and we also end up having to update more storage. 

It can work more easily when every token have their own nonces, but with all nonces handled by the account contract, you might need quite a bit of business logic to handle races etc.